contracts/cw721-base/src/msg.rs

```diff
@@ -5,12 +5,17 @@ use cosmwasm_std::Binary;
 use cw721::Expiration;
 
 #[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
-pub struct InstantiateMsg {
+pub struct InstantiateMsg<I> {
     /// Name of the NFT contract
     pub name: String,
-    /// Symbol of the NFT contract
+    /// Part of the OG ERC721 standard even though no one uses it
     pub symbol: String,
-
+    /// Universal resource identifier for this NFT Collection
+    /// Should point to a JSON file that conforms to contract level metadata
+    /// schema: https://docs.opensea.io/docs/contract-level-metadata
+    pub collection_uri: Option<String>,
```

Comment:
Isn't this just something that an implementer should put in the metadata? It assumes quite heavily that you're using NFTs in a particular way, which as noted by the sarcastic (fair) comment at line 11 is a thing that has occurred elsewhere and resulted in fields that aren't really used by all implementers 🤷

Reply:
Especially as it also has an implication on the default config response vs it just being cleanly encapsulated in the optional meta.

Reply:
> Isn't this just something that an implementer should put in the metadata?

Well `collection_uri` _is_ fairly important. It's used by `ics721` to transfer collection level metadata. Without it, a Juno NFT transferred to Stargaze would have no way to display information about the collection. For example, at the top of the [Bad Kids marketplace page](https://app.stargaze.zone/marketplace/stars19jq6mj84cnt9p7sagjxqf8hxtczwc8wlpuwe4sh62w45aheseues57n420), we show the collection image, name, and description. We use this info in other views as well, all of which wouldn't work with cross chain NFTs if they didn't have a `collection_uri`. Of course other apps using Interchain NFTs may also want to use this data. `ics721` as written currently does not support other on-chain metadata.

@shanev initially suggested we put it in the metadata extension, and it's an easy change to make, but given the spec I thought maybe it would be best to higher level than the optional metadata and always included in the default config response. Will go with what we decide as a group, but wanted to articulate my concerns and the rational for _not_ putting it in the `metadata` extension.

Reply:
I think the opensea argument isn't particularly compelling, but take the point that we probably want the same implementation as the SDK - to me, that's probably enough to swing it.

From a design point of view, I still believe that it should simply be _in_ the metadata, and that metadata is therefore simply _not optional_, since it has a key that is always required.

It strikes me as bad data structure design to have something that is metadata... not in the metadata.

This is already an issue with the opensea style URIs in the existing CW NFTs, because you end up duplicating most metadata fields _in_ the main metadata, whether you're using external JSON or on-chain metadata (which both the projects I work on use).

Reply:
I am in agreement with @the-frey here in theory. However, although `collection_uri` and `metadata` both represent top-level metadata, one is a URI and the other is an arbitrary data type that lives on-chain. You cannot represent the latter in the former, and the former in the latter works. But then its possible to have partial metadata off-chain, and partial and/or conflicting metadata on-chain. Especially considering the SDK NFT implementation, I see no harm in having both, especially since they are optional.

That said, there is a way to represent "either" in Rust. Can make this an enum like: 
```rs
pub enum Metadata<T> {
    CollectionUri(String),
    Data(T),
}
```
This is how I would prefer to do it. But the current approach is also fine imho.

Reply:
Personally, I would rather just have `metadata` and not have `collection_uri`, but unfortunately the `ics721` spec doesn't support on chain metadata. :smiling_face_with_tear:

So I am also in agreement with @the-frey in theory, but I think the spec is forcing our hand here. Maybe @ezekiiel has some context here since he's been working on an implementation of `ics721`?

Reply:
Feel like @shanev has a point that we could encapsulate part of the dilemma in an ADT and thus make it more concretely a single thing rather than two separate fields...

Kind of annoying that the ICS721 spec doesn't support on-chain metadata, as an aside.

Reply:
> Kind of annoying that the ICS721 spec doesn't support on-chain metadata, as an aside.

This is the problem with having a single thing. If we use an enum like above, any NFTs with on-chain metadata will not have collection metadata when using `ics721`.

Is this ideal? No.

Is it the end of the world to have two optional fields? Also no. :man_shrugging: